Fix FaucetFailedToBuildTx in smoke test

ffakenz opened the pull request:
<!-- Describe your change here -->

Our smoke test requires funds from the faucet to run. Once completed, it constructs a transaction to return the provided funds back to the faucet. To determine the amount of value to be returned, it subtracts a fee calculation from the remaining balance.

However, there is an issue with our fee calculation, causing the smoke test to fail with [FaucetFailedToBuildTx {reason = TxBodyErrorAdaBalanceNegative (Coin (-176))}](https://github.com/input-output-hk/hydra/actions/runs/8455143214/job/23161909806).

This PR aims to resolve the issue by removing the incorrect fee calculation and allowing the transaction to automatically balance itself during construction.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Slightly increase the hard-coded fee to return to faucet address

To work for sanchonet

ffakenz committed:
Add debug points

ffakenz committed:
Remove fee calculation

Let the makeTransactionBodyAutoBalance do its work.

ffakenz committed:
Promote it spec to prop

ch1bo commented on hydra-cluster/test/Test/Hydra/Cluster/FaucetSpec.hs:
```diff
@@ -30,27 +30,29 @@ spec = do
+            withTempDir "hydra-cluster" $ \tmpDir ->
+              withCardanoNodeDevnet (contramap FromCardanoNode tracer) tmpDir $ \node@RunningNode{networkId, nodeSocket} -> do
+                let faucetTracer = contramap FromFaucet tracer
+                actor <- generate $ elements [Alice, Bob, Carol]
```
Should: run generator using `forAll` instead of `generate`

This is now a property and we can run it more directly. This means we need to move this up before `showLogsOnFailure` (where we enter `IO`)

ch1bo commented on hydra-cluster/test/Test/Hydra/Cluster/FaucetSpec.hs:
```diff
@@ -30,27 +30,29 @@ spec = do
-            -- just be the amount of paid fees
-            difference `shouldSatisfy` (< 340_000)
+                let initialFaucetValue = selectAsset (foldMap txOutValue initialFaucetFunds) AdaAssetId
+                let finalFaucetValue = selectAsset (foldMap txOutValue finalFaucetFunds) AdaAssetId
```
Could: use `selectLovelace` here

ch1bo commented on hydra-cluster/test/Test/Hydra/Cluster/FaucetSpec.hs:
```diff
@@ -30,27 +30,29 @@ spec = do
-            -- check the faucet has one utxo extra in the end
-            length seeded `shouldBe` length remaining + 1
+                -- check the faucet has one utxo extra in the end
+                length seeded `shouldBe` length remaining + 1
```
I don't understand this assertion. (I know you have not changed it, but still).

I think the equation is not wrong here, but it does not really match the comment above? It was introduced in https://github.com/input-output-hk/hydra/commit/0cfc4e0db9b2dd3ed0113af7c3274f4fb2166bec

Checking that the faucet has one utxo more than before would be based on `initialFaucetFunds` and `finalFaucetFunds`:

```suggestion
                length finalFaucetFunds `shouldBe` length initialFaucetFunds + 1
```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -106,32 +106,24 @@ returnFundsToFaucet ::
-          Left e -> throwIO $ FaucetFailedToBuildTx{reason = e}
-          Right body -> pure body
+  buildTxBody utxo faucetAddress =
+    buildTransaction networkId nodeSocket faucetAddress utxo [] [] >>= \case
```
Could: add a comment why this if fine (i.e. explain that with no outputs a change output with all value is created and paid to the faucet address)

ch1bo reviewed (approved):
The change is fine, the test assertions are / were confusing and I left some comments to maybe improve them (as we touch this code now).

ffakenz committed:
Enhance faucet spec

- Generate the arbitrary actor within the forAll expression.
- Added an explanatory comment regarding why we do not set outputs while building the transaction.
- Make use of selectLovelace helper instead.
